# ============================================================================
# Configuration for mola::state_estimation_smoother::StateEstimationSmoother
# ============================================================================
# This file holds parameters for MOLA State Estimation Smoother that fuses
# multiple sensor inputs (IMU, odometry, GNSS) for robust pose and velocity
# estimation.
#
# Refer to:
# https://docs.mola-slam.org/latest/mola_state_estimators.html
# https://github.com/MOLAorg/mola
# ============================================================================

params:
  # --------------------------------------------------------------------------
  # Reference Frame IDs
  # --------------------------------------------------------------------------

  # Frame name for the vehicle/robot base
  # Used to publish timely pose updates:
  vehicle_frame_name: "${MOLA_TF_BASE_LINK|base_link}"

  # Reference frame for pose publication (typically 'map' or 'odom')
  # Used to publish timely pose updates.
  # See docs: https://docs.mola-slam.org/latest/mola_state_estimators.html
  reference_frame_name: "${MOLA_TF_MAP|map}"

  # The ENU geo-reference frame
  enu_frame_name: "enu"

  # --------------------------------------------------------------------------
  # Kinematic Model & Motion Factors
  # --------------------------------------------------------------------------

  # Kinematic model for internal motion model factors
  # Options: KinematicModel::ConstantVelocity, KinematicModel::Tricycle
  kinematic_model: "${MOLA_NAVSTATE_KINEMATIC_MODEL|KinematicModel::ConstantVelocity}"

  # Time window to keep past observations in the filter [seconds]
  sliding_window_length: ${MOLA_NAVSTATE_SLIDING_WINDOW_SEC|2.5}

  # Minimum time difference between frames to create a new frame [seconds]
  min_time_difference_to_create_new_frame: 0.01

  # Maximum time to extrapolate velocity model since last observation [seconds]
  # Valid estimations will be extrapolated only up to this time since the
  # last incorporated observation. If a request is done farther away, an
  # empty estimation will be returned.
  # Beyond this time, empty estimations will be returned.
  # (KITTI360 test_3 has some blackouts >=1.0 seconds while angular acceleration is high, so
  #  this threshold must be < 1.0 s for that dataset)
  max_time_to_use_velocity_model: ${MOLA_MAX_TIME_TO_USE_VELOCITY_MODEL|0.75}

  # Warning threshold: emit warning if time between keyframes exceeds this [seconds]
  # If the time between two keyframes is larger than this, a warning will be
  # emitted; but the algorithm will keep trying its best.
  time_between_frames_to_warning: 3.0

  # When adding GNSS observations, specially with consumer grade receivers with errors larger
  # than a few centimeters, we may be more permissive in the temporal distance between the GNSS
  # datum and the associated existing keyframe. This parameter is the extended, alternative value
  # to use instead of "min_time_difference_to_create_new_frame". [seconds]
  gnss_nearby_keyframe_stamp_tolerance: 1.0

  # When adding IMU observations, this is the temporal distance between the IMU reading
  # and the associated existing keyframe. This applies to gravity-oriented (IMU attitude) and
  # gravity-estimation (accelerometer) only factors, not to high-frequency IMU preintegration.
  # This parameter is the extended, alternative value
  # to use instead of "min_time_difference_to_create_new_frame". [seconds]
  imu_nearby_keyframe_stamp_tolerance: 0.10

  # Random walk model for linear acceleration uncertainty [m/s²]
  sigma_random_walk_acceleration_linear: ${MOLA_NAVSTATE_SIGMA_RANDOM_WALK_LINACC|1.0}

  # Random walk model for angular acceleration uncertainty [rad/s²]
  sigma_random_walk_acceleration_angular: ${MOLA_NAVSTATE_SIGMA_RANDOM_WALK_ANGACC|10.0}

  # Integrator uncertainty for position [m]
  sigma_integrator_position: 0.1

  # Integrator uncertainty for orientation [rad]
  sigma_integrator_orientation: 1.0

  # Uncertainty for linear velocity estimated from consecutive poses [m/s]
  sigma_twist_from_consecutive_poses_linear: 1.0

  # Uncertainty for angular velocity estimated from consecutive poses [rad/s]
  sigma_twist_from_consecutive_poses_angular: 1.0

  # If enabled, the estimator will enforce z, pitch, and roll to be always 0:
  enforce_planar_motion: ${MOLA_NAVSTATE_ENFORCE_PLANAR_MOTION|false}

  # If set, the first ever frame will also have an SE(3) edge favoring it to be the identity in
  # the "reference_frame", with a sigma given by this value. Use a small number, like 1e-6, for
  # initialing the first odometry pose near the map origin. Do not set when using geo-referenced
  # maps.
  # link_first_pose_to_reference_origin_sigma: 1e-6

  # --------------------------------------------------------------------------
  # Initial State Estimation
  # --------------------------------------------------------------------------

  # Optional initial guess for the twist (vx vy vz: m/s, wx wy wz: rad/s):
  # Initial velocity twist (linear: vx, vy, vz; angular: wx, wy, wz)
  initial_twist: ["${MOLA_INITIAL_VX|0.0}", 0.0, 0.0, 0.0, 0.0, 0.0]

  # Defaults: somewhat confident that the vehicle is near rest.
  # Change these if needed to start with the vehicle at high speed.
  # Initial linear velocity uncertainty [m/s]
  initial_twist_sigma_lin: ${MOLA_INITIAL_TWIST_SIGMA_LIN|20.0}

  # Initial angular velocity uncertainty [rad/s]
  initial_twist_sigma_ang: ${MOLA_INITIAL_TWIST_SIGMA_ANG|3.0}

  # --------------------------------------------------------------------------
  # IMU Related
  # --------------------------------------------------------------------------

  # When an IMU provides global attitude measurements (azimuth and gravity aligned), this is the
  # uncertainty or noise sigma [degrees].
  imu_attitude_sigma_deg: "${MOLA_IMU_ATTITUDE_SIGMA|2.0}"

  # When an IMU provides global attitude measurements (azimuth and gravity aligned), this must
  # define the angle (in degrees) to add to IMU yaw orientation to obtain azimuth so 0 deg is
  # North. Note that ENU axes are such vehicle yaw is 0 when pointing East instead.
  # Example cases:
  # - IMU absolute yaw=0 points True North ==> offset=0
  # - IMU absolute yaw=0 points East ==> offset=-90
  imu_attitude_azimuth_offset_deg: "${MOLA_IMU_ATTITUDE_AZIMUTH_OFFSET|0.0}"

  # When using an IMU with acceleration, use this sigma to estimate the up-vector, hence
  # gravity-align the map.
  # Set to 0 to disable.
  imu_normalized_gravity_alignment_sigma: "${MOLA_IMU_GRAVITY_ALIGNMENT_SIGMA|0.4}"

  # --------------------------------------------------------------------------
  # Geo-referencing
  # --------------------------------------------------------------------------

  # If true, this estimator will try to estimate the best geo-referencing for {enu} ->
  # {map} from incoming GNSS readings and other sensors. If false, geo-referencing is
  # assumed to be given from either these initial parameters or, if not set, from an external
  # source (e.g. a geo-referenced .mm map loaded in mola_lidar_odometry).
  estimate_geo_reference: "${MOLA_ESTIMATE_GEO_REF|false}"

  # If estimate_geo_reference is false and this is set, the geo-referencing will be taken
  # from this value and never attempted to be optimized or changed.
  # Other geo-reference information coming from external sources may override this fixed initial
  # value, though.
  # Fixed geo-reference to use when estimate_geo_reference is false
  # fixed_geo_reference: { latitude_deg: 0.0, longitude_deg: 0.0, altitude: 0.0 }

  # --------------------------------------------------------------------------
  # Nonlinear Optimization
  # --------------------------------------------------------------------------

  # Each new sensor will become a call to isam2.update(), plus this number of additional
  # refining steps. In theory, more steps lead to more accurate results.
  additional_isam2_update_steps: 3

  # --------------------------------------------------------------------------
  # Sensor Input Filtering (Regex Patterns)
  # --------------------------------------------------------------------------

  # Regular expression to filter IMU sensor labels/topics for processing
  # Default accepts all: ".*"
  # do_process_imu_labels_re: ".*"

  # Regular expression to filter odometry input labels/topics for processing
  # Default accepts all: ".*"
  # do_process_odometry_labels_re: ".*"

  # Regular expression to filter GNSS (GPS) labels/topics for processing
  # Default accepts all: ".*"
  # do_process_gnss_labels_re: ".*"
# ============================================================================
# End of Configuration
# ============================================================================
